services/wallet/history/exchange.go

```diff
@@ -0,0 +1,174 @@
+	value          float32
+	startTimestamp int64
+	endTimestamp   int64
+}
+
+// Exchange caches conversion rates in memory on a daily basis
+type Exchange struct {
+	// year map keeps a list of values with days as index in the slice for the corresponding year (key) starting from the first to the last available
+	cache map[tokenType]map[currencyType]map[yearType][]float32
+	// special case for all time information
+	allTimeCache map[tokenType]map[currencyType][]allTimeEntry
+	fetchMutex   sync.Mutex
+}
+
+func NewExchange() *Exchange {
+	return &Exchange{
+		cache: make(map[tokenType]map[currencyType]map[yearType][]float32),
+	}
+}
+
+// GetExchangeRate returns the exchange rate from token to currency in the day of the given date
+// if none exists returns "missing <element>" error
+func (e *Exchange) GetExchangeRateForDay(token tokenType, currency currencyType, date time.Time) (float32, error) {
+	e.fetchMutex.Lock()
+	defer e.fetchMutex.Unlock()
+
+	currencyMap, found := e.cache[token]
+	if !found {
+		return 0, errors.New("missing token")
+	}
+
+	yearsMap, found := currencyMap[currency]
+	if !found {
+		return 0, errors.New("missing currency")
+	}
+
+	year := date.Year()
+	valueForDays, found := yearsMap[year]
+	if !found {
+		// Search closest in all time
+		allCurrencyMap, found := e.allTimeCache[token]
+		if !found {
+			return 0, errors.New("missing token in all time data")
+		}
+
+		allYearsMap, found := allCurrencyMap[currency]
+		if !found {
+			return 0, errors.New("missing currency in all time data")
+		}
+		for _, entry := range allYearsMap {
+			if entry.startTimestamp <= date.Unix() && entry.endTimestamp > date.Unix() {
+				return entry.value, nil
+			}
+		}
+		return 0, errors.New("missing entry")
+	}
+
+	day := date.YearDay()
+	if day >= len(valueForDays) {
+		return 0, errors.New("missing day")
+	}
+	return valueForDays[day], nil
+}
+
+// fetchAndCacheRates fetches and in memory cache exchange rates for this and last year
+func (e *Exchange) FetchAndCacheMissingRates(token tokenType, currency currencyType) error {
+	// Protect REST calls also to prevent fetching the same token/currency twice
+	e.fetchMutex.Lock()
+	defer e.fetchMutex.Unlock()
+
+	// Allocate missing values
+	currencyMap, found := e.cache[token]
+	if !found {
+		currencyMap = make(map[currencyType]map[yearType][]float32)
+		e.cache[token] = currencyMap
+	}
+
+	yearsMap, found := currencyMap[currency]
+	if !found {
+		yearsMap = make(map[yearType][]float32)
+		currencyMap[currency] = yearsMap
+	}
+
+	currentTime := time.Now().UTC()
+	endOfPrevYearTime := time.Date(currentTime.Year()-1, 12, 31, 23, 0, 0, 0, time.UTC)
+
+	daysToFetch := extendDaysSliceForYear(yearsMap, endOfPrevYearTime)
+
+	curYearTime := time.Date(currentTime.Year(), currentTime.Month(), currentTime.Day(), 0, 0, 0, 0, time.UTC)
+	daysToFetch += extendDaysSliceForYear(yearsMap, curYearTime)
+	if daysToFetch == 0 {
+		return nil
+	}
+
+	onlineSource := thirdparty.NewCryptoCompare()
```

Comment:
could you inject crypocompare, rather than building a new one, it is not stateless and help to know if the api is up or down